File: ref/causal-learner/templates/explainer.py
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

from .types import Fact, Observation, Regulation, Story
from .unify import Bindings, dedup_facts, fact_entails, substitute_fact, unify_fact


EVI_WEIGHT = {
    "intervention": 1.0,
    "quasi_experiment": 0.8,
    "observational": 0.5,
}

STATUS_WEIGHT = {
    "confirmed": 1.0,
    "hypothesis": 0.85,
    "candidate": 0.65,
    "retired": 0.1,
}


@dataclass
class ExplainOptions:
    top_k: int = 5
    beam_size: int = 20
    max_depth: int = 8
    max_assumptions: int = 10

    # penalties (in log space)
    assumption_penalty: float = 1.2
    length_penalty: float = 0.05


def context_to_facts(context: Dict[str, Any]) -> List[Fact]:
    return [Fact(pred=str(k), args={}, value=v) for k, v in (context or {}).items()]


def scope_compatible(rule: Regulation, obs_context: Dict[str, Any]) -> bool:
    # simplest: all scope key-values must match
    for k, v in (rule.scope or {}).items():
        if obs_context.get(k) != v:
            return False
    return True


def rep_weight(support_n: int, counterexample_n: int) -> float:
    base = 1.0 - math.exp(-max(0, support_n) / 5.0)  # 0..1
    base = max(0.05, base)
    penalty = math.exp(-max(0, counterexample_n))
    return max(0.05, base * penalty)


def spec_weight(rule: Regulation) -> float:
    size = len(rule.pre) + len(rule.eff)
    return min(1.0, 0.4 + 0.05 * size)


def rule_score(rule: Regulation) -> float:
    w_evi = EVI_WEIGHT.get(rule.evidence_kind, 0.5)
    w_rep = rep_weight(rule.support_n, rule.counterexample_n)
    w_spec = spec_weight(rule)
    w_stat = STATUS_WEIGHT.get(rule.status, 0.8)
    score = w_evi * w_rep * w_spec * w_stat
    return max(0.01, min(1.0, score))


class EffectIndex:
    """Index from (pred, value) -> regulations that can produce it."""
    def __init__(self, regulations: List[Regulation]):
        self._idx: Dict[Tuple[str, Any], List[Regulation]] = {}
        for r in regulations:
            for eff in r.eff:
                key = (eff.pred, eff.value)
                self._idx.setdefault(key, []).append(r)

    def candidates(self, goal: Fact) -> List[Regulation]:
        return list(self._idx.get((goal.pred, goal.value), []))


@dataclass
class _Node:
    goals: List[Fact]
    regulation_ids: List[str]
    assumptions: List[Fact]
    bindings: Bindings
    score_log: float
    used: set


def _choose_goal(goals: List[Fact], known: List[Fact], bindings: Bindings, idx: EffectIndex, obs_context: Dict[str, Any]) -> int:
    """Pick the next goal index to expand: prefer goals with fewer candidate rules.

    Note: We do NOT allow a goal to be satisfied directly just because it is observed.
    Only preconditions (subgoals) are allowed to be satisfied by known facts.
    """
    best_i = 0
    best_count = 10**9
    for i, g in enumerate(goals):
        cands = [r for r in idx.candidates(g) if scope_compatible(r, obs_context)]
        count = len(cands)
        if count < best_count:
            best_count = count
            best_i = i
    return best_i


def explain_observation(obs: Observation, regulations: List[Regulation], options: Optional[ExplainOptions] = None) -> List[Story]:
    """Explain observation goals by chaining regulations backward.

    Design note:
    - obs.goals() are facts you want to *explain* (even though they are observed).
    - known facts are used only to satisfy *preconditions* during chaining.
    """
    options = options or ExplainOptions()
    goals = dedup_facts(obs.goals())

    known = dedup_facts(list(obs.facts) + context_to_facts(obs.context))
    idx = EffectIndex(regulations)

    init = _Node(
        goals=list(goals),
        regulation_ids=[],
        assumptions=[],
        bindings={},
        score_log=0.0,
        used=set(),
    )

    beam: List[_Node] = [init]
    completed: List[_Node] = []

    for depth in range(options.max_depth):
        new_beam: List[_Node] = []

        for node in beam:
            if not node.goals:
                completed.append(node)
                continue

            gi = _choose_goal(node.goals, known, node.bindings, idx, obs.context)
            g = node.goals[gi]
            rest = node.goals[:gi] + node.goals[gi+1:]

            # try expand with candidate regulations
            expanded = False
            for r in idx.candidates(g):
                if r.regulation_id in node.used:
                    continue
                if not scope_compatible(r, obs.context):
                    continue

                # unify any effect with the goal
                for eff in r.eff:
                    b2 = unify_fact(eff, g, node.bindings)
                    if b2 is None:
                        continue
                    expanded = True

                    # add preconditions as new goals, but allow them to be satisfied by known facts immediately
                    new_goals: List[Fact] = []
                    for p in r.pre:
                        p2 = substitute_fact(p, b2)
                        b_known = fact_entails(known, p2, b2)
                        if b_known is not None:
                            b2 = b_known
                            continue
                        new_goals.append(p2)

                    merged = dedup_facts(new_goals + rest)

                    rs = rule_score(r)
                    child_score = node.score_log + math.log(rs) - options.length_penalty

                    new_beam.append(_Node(
                        goals=merged,
                        regulation_ids=node.regulation_ids + [r.regulation_id],
                        assumptions=list(node.assumptions),
                        bindings=b2,
                        score_log=child_score,
                        used=set(node.used) | {r.regulation_id},
                    ))

            if expanded:
                continue

            # no rule can produce this goal: treat as assumption (abduction)
            if len(node.assumptions) < options.max_assumptions:
                new_beam.append(_Node(
                    goals=list(rest),
                    regulation_ids=list(node.regulation_ids),
                    assumptions=node.assumptions + [g],
                    bindings=dict(node.bindings),
                    score_log=node.score_log - options.assumption_penalty,
                    used=set(node.used),
                ))

        # prune beam
        new_beam.sort(key=lambda n: n.score_log, reverse=True)  # higher is better
        beam = new_beam[: options.beam_size]

    # convert completed nodes to stories
    stories: List[Story] = []
    for n in completed:
        stories.append(Story(
            regulation_ids=list(n.regulation_ids),
            assumptions=list(n.assumptions),
            score=float(n.score_log) / max(1, len(n.regulation_ids)),
        ))

    # sort and return top_k
    stories.sort(key=lambda s: s.score, reverse=True)
    return stories[: options.top_k]

File: causal-learner/mcp-server/src/core/explainer.ts
/**
 * Beam Search backward chaining explainer for the Causal Learner
 * Explains observations using causal regulation chains
 */

import type { Fact, Observation, Regulation, Story } from './types.js';
import { observationGoals } from './types.js';
import {
  Bindings,
  dedupFacts,
  factEntails,
  substituteFact,
  unifyFact,
} from './unify.js';

// Evidence kind weights
const EVI_WEIGHT: Record<string, number> = {
  intervention: 1.0,
  quasi_experiment: 0.8,
  observational: 0.5,
};

// Status weights
const STATUS_WEIGHT: Record<string, number> = {
  confirmed: 1.0,
  hypothesis: 0.85,
  candidate: 0.65,
  retired: 0.1,
};

export interface ExplainOptions {
  topK: number;
  beamSize: number;
  maxDepth: number;
  maxAssumptions: number;
  assumptionPenalty: number;
  lengthPenalty: number;
}

export const DEFAULT_EXPLAIN_OPTIONS: ExplainOptions = {
  topK: 5,
  beamSize: 20,
  maxDepth: 8,
  maxAssumptions: 10,
  assumptionPenalty: 1.2,
  lengthPenalty: 0.05,
};

function contextToFacts(context: Record<string, unknown> | undefined): Fact[] {
  if (!context) return [];
  return Object.entries(context).map(([k, v]) => ({
    pred: String(k),
    args: {},
    value: v,
  }));
}

function scopeCompatible(
  rule: Regulation,
  obsContext: Record<string, unknown> | undefined
): boolean {
  const ctx = obsContext || {};
  const scope = rule.scope || {};
  for (const [k, v] of Object.entries(scope)) {
    if (ctx[k] !== v) {
      return false;
    }
  }
  return true;
}

function repWeight(supportN: number, counterexampleN: number): number {
  const base = 1.0 - Math.exp(-Math.max(0, supportN) / 5.0);
  const baseAdjusted = Math.max(0.05, base);
  const penalty = Math.exp(-Math.max(0, counterexampleN));
  return Math.max(0.05, baseAdjusted * penalty);
}

function specWeight(rule: Regulation): number {
  const size = rule.pre.length + rule.eff.length;
  return Math.min(1.0, 0.4 + 0.05 * size);
}

export function ruleScore(rule: Regulation): number {
  const wEvi = EVI_WEIGHT[rule.evidenceKind || 'observational'] ?? 0.5;
  const wRep = repWeight(rule.supportN || 0, rule.counterexampleN || 0);
  const wSpec = specWeight(rule);
  const wStat = STATUS_WEIGHT[rule.status] ?? 0.8;
  const score = wEvi * wRep * wSpec * wStat;
  return Math.max(0.01, Math.min(1.0, score));
}

class EffectIndex {
  private idx: Map<string, Regulation[]> = new Map();

  constructor(regulations: Regulation[]) {
    for (const r of regulations) {
      for (const eff of r.eff) {
        const key = eff.pred + '|' + JSON.stringify(eff.value);
        const existing = this.idx.get(key) || [];
        existing.push(r);
        this.idx.set(key, existing);
      }
    }
  }

  candidates(goal: Fact): Regulation[] {
    const key = goal.pred + '|' + JSON.stringify(goal.value);
    return this.idx.get(key) || [];
  }
}

/** how-provenance：记录每步推导的精确路径 */
export interface ProvenanceStep {
  /** 使用的 regulation ID */
  regulationId: string;
  /** 本步要解释的 goal fact */
  matchedGoal: Fact;
  /** regulation 的哪些 pre 被已知 facts 满足 */
  satisfiedPre: Fact[];
  /** regulation 的哪些 pre 变成新的子目标 */
  newSubgoals: Fact[];
}

interface SearchNode {
  goals: Fact[];
  regulationIds: string[];
  assumptions: Fact[];
  bindings: Bindings;
  scoreLog: number;
  used: Set<string>;
  /** how-provenance 推导链 */
  provenance: ProvenanceStep[];
}

function chooseGoal(
  goals: Fact[],
  idx: EffectIndex,
  obsContext: Record<string, unknown> | undefined
): number {
  let bestI = 0;
  let bestCount = Infinity;

  for (let i = 0; i < goals.length; i++) {
    const g = goals[i];
    const cands = idx.candidates(g).filter((r) => scopeCompatible(r, obsContext));
    const count = cands.length;
    if (count < bestCount) {
      bestCount = count;
      bestI = i;
    }
  }

  return bestI;
}

export function explainObservation(
  obs: Observation,
  regulations: Regulation[],
  options?: Partial<ExplainOptions>
): Story[] {
  const opts = { ...DEFAULT_EXPLAIN_OPTIONS, ...options };
  const goals = dedupFacts(observationGoals(obs));
  const known = dedupFacts([...obs.facts, ...contextToFacts(obs.context)]);
  const idx = new EffectIndex(regulations);

  const init: SearchNode = {
    goals: [...goals],
    regulationIds: [],
    assumptions: [],
    bindings: {},
    scoreLog: 0.0,
    used: new Set(),
    provenance: [],
  };

  let beam: SearchNode[] = [init];
  const completed: SearchNode[] = [];

  for (let depth = 0; depth < opts.maxDepth; depth++) {
    const newBeam: SearchNode[] = [];

    for (const node of beam) {
      if (node.goals.length === 0) {
        completed.push(node);
        continue;
      }

      const gi = chooseGoal(node.goals, idx, obs.context);
      const g = node.goals[gi];
      const rest = [...node.goals.slice(0, gi), ...node.goals.slice(gi + 1)];

      let expanded = false;

      for (const r of idx.candidates(g)) {
        if (node.used.has(r.regulationId)) {
          continue;
        }
        if (!scopeCompatible(r, obs.context)) {
          continue;
        }

        for (const eff of r.eff) {
          let b2 = unifyFact(eff, g, node.bindings);
          if (b2 === null) {
            continue;
          }
          expanded = true;

          const newGoals: Fact[] = [];
          const satisfiedPre: Fact[] = [];
          for (const p of r.pre) {
            const p2 = substituteFact(p, b2);
            const bKnown = factEntails(known, p2, b2);
            if (bKnown !== null) {
              b2 = bKnown;
              satisfiedPre.push(p2);
              continue;
            }
            newGoals.push(p2);
          }

          const merged = dedupFacts([...newGoals, ...rest]);
          const rs = ruleScore(r);
          const childScore = node.scoreLog + Math.log(rs) - opts.lengthPenalty;

          // how-provenance：记录本步推导
          const step: ProvenanceStep = {
            regulationId: r.regulationId,
            matchedGoal: g,
            satisfiedPre,
            newSubgoals: newGoals,
          };

          newBeam.push({
            goals: merged,
            regulationIds: [...node.regulationIds, r.regulationId],
            assumptions: [...node.assumptions],
            bindings: b2,
            scoreLog: childScore,
            used: new Set([...node.used, r.regulationId]),
            provenance: [...node.provenance, step],
          });
        }
      }

      if (expanded) {
        continue;
      }

      if (node.assumptions.length < opts.maxAssumptions) {
        newBeam.push({
          goals: [...rest],
          regulationIds: [...node.regulationIds],
          assumptions: [...node.assumptions, g],
          bindings: { ...node.bindings },
          scoreLog: node.scoreLog - opts.assumptionPenalty,
          used: new Set(node.used),
          provenance: [...node.provenance],
        });
      }
    }

    newBeam.sort((a, b) => b.scoreLog - a.scoreLog);
    beam = newBeam.slice(0, opts.beamSize);
  }

  const stories: Story[] = completed.map((n) => ({
    regulationIds: [...n.regulationIds],
    assumptions: [...n.assumptions],
    score: n.scoreLog / Math.max(1, n.regulationIds.length),
    provenance: [...n.provenance],
  }));

  stories.sort((a, b) => (b.score || 0) - (a.score || 0));
  return stories.slice(0, opts.topK);
}

export { EffectIndex };
